yarpctest/recorder/recorder_test.go

```diff
@@ -0,0 +1,251 @@
+	atom := make([]byte, length)
+	for i := 0; i < length; i++ {
+		letter := r.randsrc.Intn(2 * 26)
+		if letter < 26 {
+			atom[i] = 'A' + byte(letter)
+		} else {
+			atom[i] = 'a' + byte(letter-26)
+		}
+	}
+	return string(atom)
+}
+
+// Headers returns a new randomized header.
+func (r *randomGenerator) Headers() transport.Headers {
+	headers := transport.NewHeaders()
+	size := 2 + r.randsrc.Intn(6)
+	for i := 0; i < size; i++ {
+		headers = headers.With(r.Atom(), r.Atom())
+	}
+	return headers
+}
+
+// Request returns a new randomized request.
+func (r *randomGenerator) Request() (transport.Request, []byte) {
+	bodyData := []byte(r.Atom())
+
+	return transport.Request{
+		Caller:          r.Atom(),
+		Service:         r.Atom(),
+		Encoding:        transport.Encoding(r.Atom()),
+		Procedure:       r.Atom(),
+		Headers:         r.Headers(),
+		ShardKey:        r.Atom(),
+		RoutingKey:      r.Atom(),
+		RoutingDelegate: r.Atom(),
+		Body:            ioutil.NopCloser(bytes.NewReader(bodyData)),
+	}, bodyData
+}
+
+func TestHash(t *testing.T) {
+	rgen := newRandomGenerator(42)
+	request, body := rgen.Request()
+
+	recorder := NewRecorder(t)
+	referenceHash := recorder.hashRequest(&request, body)
+
+	require.Equal(t, "7195d5a712201d2a", referenceHash)
+
+	// Caller
+	r := request
+	r.Caller = rgen.Atom()
+	assert.NotEqual(t, recorder.hashRequest(&r, body), referenceHash)
+
+	// Service
+	r = request
+	r.Service = rgen.Atom()
+	assert.NotEqual(t, recorder.hashRequest(&r, body), referenceHash)
+
+	// Encoding
+	r = request
+	r.Encoding = transport.Encoding(rgen.Atom())
+	assert.NotEqual(t, recorder.hashRequest(&r, body), referenceHash)
+
+	// Procedure
+	r = request
+	r.Procedure = rgen.Atom()
+	assert.NotEqual(t, recorder.hashRequest(&r, body), referenceHash)
+
+	// Headers
+	r = request
+	r.Headers = rgen.Headers()
+	assert.NotEqual(t, recorder.hashRequest(&r, body), referenceHash)
+
+	// ShardKey
+	r = request
+	r.ShardKey = rgen.Atom()
+	assert.NotEqual(t, recorder.hashRequest(&r, body), referenceHash)
+
+	// RoutingKey
+	r = request
+	r.RoutingKey = rgen.Atom()
+	assert.NotEqual(t, recorder.hashRequest(&r, body), referenceHash)
+
+	// RoutingDelegate
+	r = request
+	r.RoutingDelegate = rgen.Atom()
+	assert.NotEqual(t, recorder.hashRequest(&r, body), referenceHash)
+
+	// Body
+	r = request
+	b := []byte(rgen.Atom())
+	assert.NotEqual(t, recorder.hashRequest(&r, b), referenceHash)
+}
+
+var testingTMockFatal = struct{}{}
+
+type testingTMock struct {
+	*testing.T
+
+	fatalCount int
+}
+
+func (t *testingTMock) Fatal(args ...interface{}) {
+	t.Logf("counting fatal: %s", args)
+	t.fatalCount++
+	panic(testingTMockFatal)
+}
+
+func TestOverwriteReplay(t *testing.T) {
+	tMock := testingTMock{t, 0}
+
+	dir, err := ioutil.TempDir("", "yarpcgorecorder")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir) // clean up
+
+	func() {
+		// First we double check that our cache is empty.
+		recorder := NewRecorder(&tMock, Option{
```

Comment:
This test is a bit of a jumble. Can you break this apart? Pull out client and
server set up logic and then:

```
TestReplay
// Either write a hard-coded YAML into the temporary directory or have a YAML
// file checked into the repo that you will try to replay in this test. Don't
// Start() the server in this test.

TestReplayMissingRecord
// This is basically this first func() { .. } you have. Don't Start() the
// server.

TestOverwrite
// Either write a hard-coded YAML into the temporary directory or copy a YAML
// that you checked into the repo into that directory, record over it with a
// response from the server that is not what is in the YAML file, and verify
// that the contents of the two files don't match.

TestAppendAndReplay
// Have some existing records, record new ones, and replay.
```